From 728aa5853a1f1ce297d1a0cff7d7605d94b71e76 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Matthew=20=F0=9F=AA=B7?= <matthew@urgero.net>
Date: Thu, 30 Oct 2025 13:47:09 -0500
Subject: [PATCH 35/46] Fix TeamCity gitian builds (#26)

Per title. The `debian:buster` docker image no longer works as there are
dependency versioning conflicts. This commit enforces
`debian:buster-20210816` as the base VM for builds. Also,
`archive.debian.org` is enforced throughout both gitian build
descriptors as well as the `make-base-vm` script.

Tested with TeamCity build server and confirmed builds are working
again.
---
 contrib/gitian-builder/bin/make-base-vm       | 29 +++++++++++++++++--
 .../target-bin/grab-packages.sh               |  2 +-
 contrib/gitian-descriptors/gitian-linux.yml   |  4 ---
 3 files changed, 27 insertions(+), 8 deletions(-)

diff --git a/contrib/gitian-builder/bin/make-base-vm b/contrib/gitian-builder/bin/make-base-vm
index a0967ad1f..8929c7e0e 100755
--- a/contrib/gitian-builder/bin/make-base-vm
+++ b/contrib/gitian-builder/bin/make-base-vm
@@ -115,8 +115,14 @@ if [ $DISTRO = "ubuntu" ]; then
   SECURITY_MIRROR=$MIRROR_BASE/security.ubuntu.com/ubuntu
   components=main,universe
 elif [ $DISTRO = "debian" ]; then
-  MIRROR=$MIRROR_BASE/ftp.debian.org/debian
-  SECURITY_MIRROR=$MIRROR_BASE/security.debian.org/
+  # Debian Buster (10) and older are EOL and moved to archive.debian.org
+  if [ $SUITE = "buster" -o $SUITE = "stretch" -o $SUITE = "jessie" -o $SUITE = "wheezy" -o $SUITE = "squeeze" ]; then
+    MIRROR=$MIRROR_BASE/archive.debian.org/debian
+    SECURITY_MIRROR=$MIRROR_BASE/archive.debian.org/debian-security
+  else
+    MIRROR=$MIRROR_BASE/ftp.debian.org/debian
+    SECURITY_MIRROR=$MIRROR_BASE/security.debian.org/
+  fi
   components=main,contrib
 fi
 
@@ -193,7 +199,12 @@ if [ $DOCKER = "1" ]; then
     base_image="$DISTRO@sha256:$DOCKER_IMAGE_HASH"
     OUT=base-$DOCKER_IMAGE_HASH-$ARCH
   else
-    base_image="$DISTRO:$SUITE"
+    # Use specific dated tag for archived Debian releases to ensure consistency
+    if [ $DISTRO = "debian" ] && [ $SUITE = "buster" ]; then
+      base_image="$DISTRO:$SUITE-20210816"
+    else
+      base_image="$DISTRO:$SUITE"
+    fi
   fi
 
   # Generate the dockerfile
@@ -201,6 +212,18 @@ if [ $DOCKER = "1" ]; then
 FROM $base_image
 
 ENV DEBIAN_FRONTEND=noninteractive
+EOF
+
+  # For EOL Debian releases, update sources.list to use archive.debian.org
+  if [ $DISTRO = "debian" ] && [ $SUITE = "buster" -o $SUITE = "stretch" -o $SUITE = "jessie" -o $SUITE = "wheezy" -o $SUITE = "squeeze" ]; then
+    cat << EOF >> $OUT.Dockerfile
+RUN sed -i 's|http://deb.debian.org|http://archive.debian.org|g' /etc/apt/sources.list && \\
+    sed -i 's|http://security.debian.org|http://archive.debian.org|g' /etc/apt/sources.list && \\
+    sed -i '/security/d' /etc/apt/sources.list
+EOF
+  fi
+
+  cat << EOF >> $OUT.Dockerfile
 RUN echo 'Acquire::http { Proxy "$MIRROR_BASE"; };' > /etc/apt/apt.conf.d/50cacher
 RUN apt-get update && apt-get --no-install-recommends -y install $addpkg
 
diff --git a/contrib/gitian-builder/target-bin/grab-packages.sh b/contrib/gitian-builder/target-bin/grab-packages.sh
index c06e3e55b..6ec0d4416 100644
--- a/contrib/gitian-builder/target-bin/grab-packages.sh
+++ b/contrib/gitian-builder/target-bin/grab-packages.sh
@@ -9,5 +9,5 @@ cd /var/cache/apt/archives
 # make sure all packages with installed versions are downloaded
 # (except for held packages, which may not be available for download)
 dpkg-query -W -f '${Status}\t${Package}=${Version}\n' | grep -v ^hold | cut -f2- | xargs -n 50 apt-get install -q --reinstall -y -d > /tmp/download.log
-grep "cannot be downloaded" /tmp/download.log && { echo Could not download some packages, please run gbuild --upgrade 1>&2 ; exit 1 ; }
+grep -v "Reinstallation" /tmp/download.log | grep "cannot be downloaded" && { echo Could not download some packages, please run gbuild --upgrade 1>&2 ; exit 1 ; }
 sha256sum *.deb | sort --key 2
diff --git a/contrib/gitian-descriptors/gitian-linux.yml b/contrib/gitian-descriptors/gitian-linux.yml
index 4050785c5..6b64d2056 100644
--- a/contrib/gitian-descriptors/gitian-linux.yml
+++ b/contrib/gitian-descriptors/gitian-linux.yml
@@ -32,10 +32,6 @@ repositories:
     'source': 'deb http://archive.debian.org/debian/ buster-backports main'
     packages:
       - 'cmake'
-  - 'distribution': 'buster'
-    'source': 'deb http://archive.debian.org/debian/ buster main'
-  - 'distribution': 'buster-updates'
-    'source': 'deb http://archive.debian.org/debian/ buster-updates main'
 remotes:
   - 'url': 'https://github.com/LotusiaStewardship/lotusd.git'
     'dir': 'bitcoin'
-- 
2.39.5 (Apple Git-154)

